Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,189 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    LinearMap.range l12 ≤ LinearMap.ker l23 := by
+  -- let n1 ∈ N1, we need to show l23 (l12 n1) = 0, suppose this is not the case
+  rintro _ ⟨n1, rfl⟩
+  show l23 (l12 n1) = 0
+  by_contra! hn1
+  -- denote `E` as the submodule spanned by `l23 (l12 n1)`, then by `l23 (l12 n1) ≠ 0`, we have
+  -- `E ≠ 0`
+  let E : Submodule R N3 := Submodule.span R {l23 (l12 n1)}
+  have hE : Nontrivial E :=
+    ⟨0, ⟨⟨l23 (l12 n1), Submodule.mem_span_singleton_self _⟩, Subtype.coe_ne_coe.1 hn1.symm⟩⟩
+
+  -- since `N1 ⊗ M -> N2 ⊗ M -> N3 ⊗ M` is exact, we have `l23 (l12 n1) ⊗ₜ[R] m = 0` for all `m : M`
+  have eq1 : ∀ (m : M), l23 (l12 n1) ⊗ₜ[R] m = 0 := fun m ↦
+    ex.apply_apply_eq_zero (n1 ⊗ₜ[R] m)
+  -- Then `E ⊗ M = 0` becase:
+  have eq0 : (⊤ : Submodule R (E ⊗[R] M)) = ⊥ := by
+    -- suppose `x ∈ E ⊗ M`, we aim to show `x = 0`
+    ext x
+    simp only [Submodule.mem_top, Submodule.mem_bot, true_iff]
+    have mem : x ∈ (⊤ : Submodule R _) := ⟨⟩
+    rw [← TensorProduct.span_tmul_eq_top, mem_span_set] at mem
+    obtain ⟨c, hc, rfl⟩ := mem
+    choose b a hy using hc
+    let r :  ⦃a : E ⊗[R] M⦄ → a ∈ ↑c.support → R := fun a ha =>
+      Submodule.mem_span_singleton.1 (b ha).2 |>.choose
+    have hr : ∀ ⦃i : E ⊗[R] M⦄ (hi : i ∈ c.support), b hi =
+        r hi • ⟨l23 (l12 n1), Submodule.mem_span_singleton_self _⟩ := by
+      intro i hi
+      ext
+      exact Submodule.mem_span_singleton.1 (b hi).2 |>.choose_spec.symm
+    -- Since `x ∈ E ⊗ M`, since `M` is flat and `E -> N1` is injective, we only need to check the
+    -- equality in `N1 ⊗ M`. we write `x = ∑ μᵢ • (l23 (l12 n1)) ⊗ mᵢ = ∑ μᵢ • 0 = 0`
+    -- (remember `E = span {l23 (l12 n1)}` and `eq1`)
+    refine Finset.sum_eq_zero fun i hi => show c i • i = 0 from
+      (Module.Flat.rTensor_preserves_injective_linearMap (M := M) E.subtype <|
+              Submodule.injective_subtype E) ?_
+    rw [← hy hi, hr hi, smul_tmul, map_smul, LinearMap.rTensor_tmul, Submodule.subtype_apply, eq1,
+      smul_zero, map_zero]
+  have : Subsingleton (E ⊗[R] M) := subsingleton_iff_forall_eq 0 |>.2 fun x =>
+    show x ∈ (⊥ : Submodule R _) from eq0 ▸ ⟨⟩
+
+  -- but `E ⊗ M = 0` implies `E = 0` because `M` is faithfully flat and this is a contradiction.
+  exact not_subsingleton_iff_nontrivial.2 inferInstance <| fl.rTensor_reflects_triviality R M E
+
+lemma rTensor_reflects_exact [fl : FaithfullyFlat R M]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    Function.Exact l12 l23 := LinearMap.exact_iff.2 <| by
+  have complex : LinearMap.range l12 ≤ LinearMap.ker l23 := range_le_ker_of_exact_rTensor R M _ _ ex
+  -- by previous lemma, we need only to show `ker l23 ≤ range l12`.
+  refine le_antisymm ?_ complex
+  -- We consider the cohomology of `N1 -> N2 -> N3`, i.e. let `H := ker l23 ⧸ range l12`. It is
+  -- enough to show `H = 0` because `H = 0` implies `x ∈ range l12`.
```

Comment:
This is a bit weird. The previous lemma showed range <= ker, and hence H makes sense. But I would not say that H = 0 implies ker <= range; I would say that H = 0 implies ker = range. How about

-- By the previous lemma we have that `range l12 ≤ ker l23` and hence the quotient `H := ker l23 ⧸ range l12`
-- makes sense. Hence our goal `ker l23 = range l12` follows from the claim that `H = 0`.

And do you need to `refine le_antisymm ?_ complex`?

Reply:
Thanks.

> And do you need to refine le_antisymm ?_ complex?

No. Thanks